dbcon/joblist/diskjoinstep.cpp

```diff
@@ -527,49 +560,117 @@ void DiskJoinStep::joinFcn()
   {
     reportStats();
     outputDL->endOfInput();
     closedOutput = true;
   }
 }
 
+void DiskJoinStep::initializeFIFO(const uint32_t threadCount)
+{
+  loadFIFO.clear();
+  buildFIFO.clear();
+
+  for (uint32_t i = 0; i < threadCount; ++i)
+  {
+    boost::shared_ptr<joblist::FIFO<boost::shared_ptr<LoaderOutput>>> lFIFO(
+        new FIFO<boost::shared_ptr<LoaderOutput>>(1, 1));
+    boost::shared_ptr<joblist::FIFO<boost::shared_ptr<BuilderOutput>>> bFIFO(
+        new FIFO<boost::shared_ptr<BuilderOutput>>(1, 1));
+
+    loadFIFO.push_back(lFIFO);
+    buildFIFO.push_back(bFIFO);
+  }
+}
+
+void DiskJoinStep::processJoinPartitions(const uint32_t threadID, const uint32_t smallSideSizeLimitPerThread,
+                                         const std::vector<JoinPartition*>& joinPartitions)
+{
+  std::vector<uint64_t> pipelineThreads;
+  pipelineThreads.reserve(3);
+  pipelineThreads.push_back(
+      jobstepThreadPool.invoke(Loader(this, threadID, smallSideSizeLimitPerThread, joinPartitions)));
+  pipelineThreads.push_back(jobstepThreadPool.invoke(Builder(this, threadID)));
+  pipelineThreads.push_back(jobstepThreadPool.invoke(Joiner(this, threadID)));
+  jobstepThreadPool.join(pipelineThreads);
+}
+
+void DiskJoinStep::prepareJobs(const std::vector<JoinPartition*>& joinPartitions,
+                               std::vector<std::vector<JoinPartition*>>& joinPartitionsJobs)
+{
+  const uint32_t issuedThreads = jobstepThreadPool.getIssuedThreads();
+  const uint32_t maxNumOfThreads = jobstepThreadPool.getMaxThreads();
+  const uint32_t numOfThreads =
+      std::min(std::min(maxNumOfJoinThreads, std::max(maxNumOfThreads - issuedThreads, (uint32_t)1)),
+               (uint32_t)joinPartitions.size());
+  const uint32_t workSize = joinPartitions.size() / numOfThreads;
+
+  uint32_t offset = 0;
+  joinPartitionsJobs.reserve(numOfThreads);
+  for (uint32_t threadNum = 0; threadNum < numOfThreads; ++threadNum, offset += workSize)
+  {
+    auto start = joinPartitions.begin() + offset;
+    auto end = start + workSize;
+    std::vector<JoinPartition*> joinPartitionJob(start, end);
+    joinPartitionsJobs.push_back(std::move(joinPartitionJob));
+  }
+
+  for (uint32_t i = 0, e = joinPartitions.size() % numOfThreads; i < e; ++i, ++offset)
+    joinPartitionsJobs[i].push_back(joinPartitions[offset]);
+}
+
+void DiskJoinStep::outputResult(const std::vector<rowgroup::RGData>& result)
+{
+  std::lock_guard<std::mutex> lk(outputMutex);
+  for (const auto &rgData : result)
+    outputDL->insert(rgData);
+}
+
+void DiskJoinStep::spawnJobs(const std::vector<std::vector<JoinPartition*>>& joinPartitionsJobs,
+                             const uint32_t smallSideSizeLimitPerThread)
+{
+  const uint32_t threadsCount = joinPartitionsJobs.size();
+  std::vector<uint64_t> processorThreadsId;
+  processorThreadsId.reserve(threadsCount);
+  for (uint32_t threadID = 0; threadID < threadsCount; ++threadID)
+    processorThreadsId.push_back(jobstepThreadPool.invoke(
```

Comment:
It is worth to add parenthesis here to avoid potential issues if the scope will be extended in future.